New Addon - Fullscreen button on left

Norbiros opened the pull request:
<!-- Which issue(s) does this pull request fix or resolve? -->

Resolves #3937
Updates this PR: #2010

### Changes

Moves feature that [TheColaber](https://github.com/ScratchAddons/ScratchAddons/pull/2010#issuecomment-821754505) proposed to new addon.

### Tests

Everything works great - Firefox 95.
![Screenshot from 2022-05-30 18-51-04](https://user-images.githubusercontent.com/89837724/171034254-bc8718ba-9e18-4b22-ae49-f4647d7e9a80.png)


### Other

Apparently Github closes PR on fast forward, so I need to create second PR (previous #4621)
Also, if anyone has better suggestions to name/description, please suggest!

TheColaber commented:
This is great! 
![image](https://user-images.githubusercontent.com/72760579/171323434-8ceb61ad-fced-4cad-900f-216529d781a5.png)Just 2 things that need to be fixed first...
1. the turbo mode icon should be in front of the text.
2. there is no spacing for the debugger icon, since the debugger had a margin-right, and didn't account for this. Since I'm not a CSS expert, I'm not sure what the best fix for this is.

TheColaber commented on addons/editor-buttons-reverse-order/addon.json:
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "Fullscreen button on left",
+  "description": "Moves fullscreen button to the left, and green flag, pause and stop button to the right.",
```
@WorldLanguages Help us out with a name and description.

Secret-chest replied:
Name: Flipped stage control order
Description: Same

mxmou replied:
I suggest "Reverse order of project controls" (the green flag and stop sign aren't stage controls). There should also be "as in Scratch 2.0" or something similar at the end of the description.

Norbiros replied:
I think that mxmou's idea for name is the best, but what with description?
Maybe something like that: This addon changes order of project controls. You can use this addon to recreate Scratch 2.0.
I don't think that we should add more information...

mxmou replied:
Descriptions should mention everything an addon does to make it appear when searching for those things.

Norbiros replied:
This addon moves green flag, stop and pause button to the right, and other elements to the left. It changes order of those elements as in Scratch 2.0. We suggest to use this addon with Display stage on left side.
Something like that?

mxmou commented:
I don't think it makes sense to reverse the order of the stage size controls.

mxmou commented on addons/editor-buttons-reverse-order/userstyle.css:
```diff
@@ -50,10 +50,13 @@
+.sa-debugger-container {
+  margin-left: 0.2rem;
+  margin-right: 0px !important;
 }
```
Use separate rules for `[dir="ltr"]` and `[dir="rtl"]`. You can also move the addon to the "themes" section at the bottom of `addons.json` to make the `!important` unnecessary.

mxmou replied:
I reverted the change to djdolphin's code because it worked fine before. I also made this compatible with the gamepad addon.

Norbiros commented:
@mxmou, you didn't fix all things with that commit. In rtl menu looks like this:
![Screenshot from 2022-06-01 20-56-02](https://user-images.githubusercontent.com/89837724/171480718-513c15ee-f8ec-47e6-97a2-dd19b26354d9.png)
Also, I don't like idea of adding addon things to css, what with new addons (someone gonna create addon and forget about adding it to this list, and after release users will report bug with this)?

So I propose my last PR, or this:
```css
[class*="stage-header_stage-menu-wrapper"],
[class*="stage-header_stage-size-row"],
[class*="stage-header_stage-size-toggle-group"] {
  flex-direction: row-reverse;
}

[class*="stage-header_stage-size-toggle-group"] {
  margin: 0 0.2rem;
}


[dir="ltr"] [class*="stage-header_stage-button-first"] {
  border-top-right-radius: calc(0.5rem / 2);
  border-bottom-right-radius: calc(0.5rem / 2);

  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

[dir="ltr"] [class*="stage-header_stage-button-last"] {
  border-left: 1px solid hsla(0, 0%, 0%, 0.15);
  border-top-left-radius: calc(0.5rem / 2);
  border-bottom-left-radius: calc(0.5rem / 2);

  border-right: none;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

[dir="rtl"] [class*="stage-header_stage-button-last"] {
  border-top-left-radius: calc(0.5rem / 2);
  border-bottom-left-radius: calc(0.5rem / 2);

  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

[dir="rtl"] [class*="stage-header_stage-button-first"] {
  border-right: 1px solid hsla(0, 0%, 0%, 0.15);
  border-top-right-radius: calc(0.5rem / 2);
  border-bottom-right-radius: calc(0.5rem / 2);

  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

[class*="green-flag_green-flag_"],
[class*="stop-all_stop-all_"],
.pause-btn {
  order: 1;
}
```
With this solution you need to remove this from debbuger's userstyle:
```css
[dir="rtl"] .sa-debugger-container {
  margin-left: 0.2rem;
}
```

mxmou commented:
> @mxmou, you didn't fix all things with that commit. In rtl menu looks like this: ![Screenshot from 2022-06-01 20-56-02](https://user-images.githubusercontent.com/89837724/171480718-513c15ee-f8ec-47e6-97a2-dd19b26354d9.png)

Not sure what could cause that. It looks as expected for me when I set the language to Arabic:
![image](https://user-images.githubusercontent.com/51849865/171485872-9ce47fb2-1e4e-470f-a5a3-fffff562fc2e.png)

> Also, I don't like idea of adding addon things to css, what with new addons (someone gonna create addon and forget about adding it to this list, and after release users will report bug with this)?

It's difficult to make addons compatible with each other without adding CSS specific to certain addons. Your solution will not work when there is more than one button (debugger and gamepad).

Norbiros commented:
> > @mxmou, you didn't fix all things with that commit. In rtl menu looks like this: ![Screenshot from 2022-06-01 20-56-02](https://user-images.githubusercontent.com/89837724/171480718-513c15ee-f8ec-47e6-97a2-dd19b26354d9.png)
> 
> Not sure what could cause that. It looks as expected for me when I set the language to Arabic: ![image](https://user-images.githubusercontent.com/51849865/171485872-9ce47fb2-1e4e-470f-a5a3-fffff562fc2e.png)
> 
Because you didn't turn on this addon?
> > Also, I don't like idea of adding addon things to css, what with new addons (someone gonna create addon and forget about adding it to this list, and after release users will report bug with this)?
> 
> It's difficult to make addons compatible with each other without adding CSS specific to certain addons. Your solution will not work when there is more than one button (debugger and gamepad).

It works (I'm talking about posted solution, not commit)?
![Screenshot from 2022-06-01 22-12-35](https://user-images.githubusercontent.com/89837724/171492646-f14b0625-13a1-46c8-9a08-ef65999f72c6.png)

mxmou commented:
> > Not sure what could cause that. It looks as expected for me when I set the language to Arabic
> 
> Because you didn't turn on this addon? (green flag is on left side)

It's supposed to be on the left side in RTL.

Norbiros commented:
> > > Not sure what could cause that. It looks as expected for me when I set the language to Arabic
> > 
> > 
> > Because you didn't turn on this addon? (green flag is on left side)
> 
> It's supposed to be on the left side in RTL.

Oh okay... So I don't know how that made that bug... Sorry for that. 
So now this PR is ready to merge (after your commit)? We need to do/fix something else?

mxmou commented:
> Oh okay... So I don't know how that made that bug... Sorry for that. So now this PR is ready to merge (after your commit)? We need to do/fix something else?

I think we still need a better name and description. I also suggested not changing the order of the small and large stage buttons.

Norbiros commented:
Okay, with name and desc we ae waiting for WL... I agree with not changing order of those elements, This code is good (I want to post it, not create commit that you gonna reverse)?
```css
[class*="stage-header_stage-menu-wrapper"],
[class*="stage-header_stage-size-row"] {
  flex-direction: row-reverse;
}

[dir="ltr"] [class*="stage-header_stage-size-toggle-group"],
[dir="ltr"] .sa-debugger-container,
[dir="ltr"] .sa-gamepad-container {
  margin-right: 0;
  margin-left: 0.2rem;
}

[dir="rtl"] [class*="stage-header_stage-size-toggle-group"],
[dir="rtl"] .sa-debugger-container,
[dir="rtl"] .sa-gamepad-container {
  margin-left: 0;
  margin-right: 0.2rem;
}

[dir="ltr"] [class*="stage-header_stage-button-first"] {
  border-left: 1px solid hsla(0, 0%, 0%, 0.15);
  border-top-left-radius: calc(0.5rem / 2);
  border-bottom-left-radius: calc(0.5rem / 2);

  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

[dir="ltr"] [class*="stage-header_stage-button-last"] {
  border-top-right-radius: calc(0.5rem / 2);
  border-bottom-right-radius: calc(0.5rem / 2);

  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

[dir="rtl"] [class*="stage-header_stage-button-first"] {
  border-right: 1px solid hsla(0, 0%, 0%, 0.15);
  border-top-right-radius: calc(0.5rem / 2);
  border-bottom-right-radius: calc(0.5rem / 2);

  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

[dir="rtl"] [class*="stage-header_stage-button-last"] {
  border-top-left-radius: calc(0.5rem / 2);
  border-bottom-left-radius: calc(0.5rem / 2);

  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}


[class*="green-flag_green-flag_"],
[class*="stop-all_stop-all_"],
.pause-btn {
  order: 1;
}
```
I just removed row reverse, and changed where round corner should be, because it is different now...

mxmou commented:
@Norbiros That looks good. You can just remove the code that changes rounded corners and borders because it's now redundant.

Norbiros commented:
Okay. Now I don't know if this addon should have djophin in credits, we have just 6 lines from his css... I think no credits in this addon is the best option, because it is very small.

DatGamerYoloYT commented:
les gooooooooo

WorldLanguages commented:
Beautiful CSS solution!

One question tho, if there's a new addon that also uses the `stageHeader` shared space (along with debugger and gamepad), will we need to modify this addon's source to account for that? Maybe we can find better selectors than harcoding `.sa-debugger-container` and `.sa-gamepad-container`?

WorldLanguages commented:
Let's also see if we can find suitable addon name and description for this (it's easy to show a screenshot of this feature, but harder to explain it with words, tbh).

The description of the `columns` addons also wants to state it's similar to 2.0, let's be consistent: `like in Scratch 2.0`
So the resulting description would be:
`Moves the green flag and stop buttons to the right and the full screen button to the left, like in Scratch 2.0.`
(Description must end with a dot)

About the addon name, `Green flag and stop buttons on the right` would be better in a way (no need to read the description with an addon name like that), but I don't think it's better than `Reverse order of project controls`.

WorldLanguages commented on addons/editor-buttons-reverse-order/addon.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "Reverse order of project controls",
+  "description": "Moves the green flag and stop buttons to the right and the full screen button to the left, as in Scratch 2.0",
+  "tags": ["editor", "projectPlayer"],
```
I think we should add the featured tag. Other 2.0-like addons have it (columns, editor-stage-left, etc.)

mxmou commented:
> One question tho, if there's a new addon that also uses the `stageHeader` shared space (along with debugger and gamepad), will we need to modify this addon's source to account for that? Maybe we can find better selectors than hardcoding `.sa-debugger-container` and `.sa-gamepad-container`?

We don't know if elements added to the stage header by future addons will have the same margin as the existing ones so I think it makes sense to style them separately. It's possible to set `direction: rtl` instead of `flex-direction: row-reverse` to change the margins automatically but that's a bit too hacky.

Norbiros commented:
> Beautiful CSS solution!
> 
> One question tho, if there's a new addon that also uses the `stageHeader` shared space (along with debugger and gamepad), will we need to modify this addon's source to account for that? Maybe we can find better selectors than hardcoding `.sa-debugger-container` and `.sa-gamepad-container`?

Maybe we can change this:
```js
[dir="rtl"] [class*="stage-header_stage-size-toggle-group"],
[dir="rtl"] .sa-debugger-container,
[dir="rtl"] .sa-gamepad-container {
  margin-left: 0;
  margin-right: 0.2rem;
}
```
to this:
```js
[dir="ltr"] [class*="stage-header_stage-size-toggle-group"],
[dir="ltr"] [class*="sa"] {
  margin-right: 0;
  margin-left: 0.2rem;
}
```
*The same for rtl*.
